### PR TITLE
Kyle0654 inpaint improvement - with refactoring from PR #1221

### PR DESCRIPTION
**Pull request by lstein**
# I made a few suggested modifications to your PR:
    
1. pass a PIL.Image to img2img and inpaint rather than tensor. This avoids passing two extra arguments and cleans up generate() (a bit)
2. Clipseg generates an "L" or "1" format mask, so the mask won't necessarily have any transparency. Instead of getting the mask's "A" channel, just convert it to "L".
3. Moved the postprocessing code into an overridden sample_to_image() method, keeping the generation code separate from the image processing code.

BTW, the CLI tries to autogenerate masks from images that contain transparency so that you don't have to pass both the original and the transparent version. When using this with your unmodified PR, I got a crash:

```
 ╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ /u/lstein/projects/SD/InvokeAI/./scripts/invoke.py:855 in <module>                               │
│                                                                                                  │
│   852 ######################################                                                     │
│   853                                                                                            │
│   854 if __name__ == '__main__':                                                                 │
│ ❱ 855 │   main()                                                                                 │
│   856                                                                                            │
│                                                                                                  │
│ /u/lstein/projects/SD/InvokeAI/./scripts/invoke.py:104 in main                                   │
│                                                                                                  │
│   101 │   │   )                                                                                  │
│   102 │                                                                                          │
│   103 │   try:                                                                                   │
│ ❱ 104 │   │   main_loop(gen, opt)                                                                │
│   105 │   except KeyboardInterrupt:                                                              │
│   106 │   │   print("\ngoodbye!")                                                                │
│   107                                                                                            │
│                                                                                                  │
│ /u/lstein/projects/SD/InvokeAI/./scripts/invoke.py:330 in main_loop                              │
│                                                                                                  │
│   327 │   │   │   if operation == 'generate':                                                    │
│   328 │   │   │   │   catch_ctrl_c = infile is None # if running interactively, we catch keybo   │
│   329 │   │   │   │   opt.last_operation='generate'                                              │
│ ❱ 330 │   │   │   │   gen.prompt2image(                                                          │
│   331 │   │   │   │   │   image_callback=image_writer,                                           │
│   332 │   │   │   │   │   step_callback=step_callback,                                           │
│   333 │   │   │   │   │   catch_interrupts=catch_ctrl_c,                                         │
│                                                                                                  │
│ /u/lstein/projects/SD/InvokeAI/ldm/generate.py:396 in prompt2image                               │
│                                                                                                  │
│    393 │   │   │   │   log_tokens    =self.log_tokenization                                      │
│    394 │   │   │   )                                                                             │
│    395 │   │   │                                                                                 │
│ ❱  396 │   │   │   init_image,mask_image,pil_image,pil_mask = self._make_images(                 │
│    397 │   │   │   │   init_img,                                                                 │
│    398 │   │   │   │   init_mask,                                                                │
│    399 │   │   │   │   width,                                                                    │
│                                                                                                  │
│ /u/lstein/projects/SD/InvokeAI/ldm/generate.py:655 in _make_images                               │
│                                                                                                  │
│    652 │   │   elif text_mask:                                                                   │
│    653 │   │   │   init_mask = self._txt2mask(image, text_mask, width, height, fit=fit)          │
│    654 │   │                                                                                     │
│ ❱  655 │   │   return init_image, init_mask, image, mask_image                                   │
│    656 │                                                                                         │
│    657 │   def _make_base(self):                                                                 │
│    658 │   │   if not self.generators.get('base'):                                               │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
```

### CODE REVIEW

**Comment by lstein** [comment]
I'm afraid some other stuff got in there because of divergence between your fork and mine. The only files that changed are:
1. ldm/generate.py
2. ldm/invoke/generator/img2img.py
3. ldm/invoke/generator/inpaint.py

File: ldm/invoke/generator/inpaint.py [hunk 1]
```diff
@@ -96,30 +104,46 @@ def make_image(x_T):
...
+        gen_result = super().sample_to_image(samples).convert('RGB')
+        
+        pil_mask = self.pil_mask
+        pil_image = self.pil_image
+        mask_blur_radius = self.mask_blur_radius
 
-            # Get numpy version
-            np_gen_result = np.asarray(gen_result, dtype=np.uint8)
+        # Get the original alpha channel of the mask
+        pil_init_mask = pil_mask.convert('L')
```
Comment (Kyle0654) [inline ldm/invoke/generator/inpaint.py:119]: The UI is currently passing RGBA as the mask. Could call it a bug I guess, and make the UI fix it.

File: ldm/invoke/generator/inpaint.py [hunk 2]
```diff
@@ -31,17 +35,21 @@ def get_make_image(self,prompt,sampler,steps,cfg_scale,ddim_eta,
...
-        pil_init_image = pil_image.convert('RGBA') # Add an alpha channel if one doesn't exist
-
-        # Build an image with only visible pixels from source to use as reference for color-matching.
-        # Note that this doesn't use the mask, which would exclude some source image pixels from the
-        # histogram and cause slight color changes.
-        init_rgb_pixels = np.asarray(pil_image.convert('RGB'), dtype=np.uint8).reshape(pil_image.width * pil_image.height, 3)
-        init_a_pixels = np.asarray(pil_init_image.getchannel('A'), dtype=np.uint8).reshape(pil_init_mask.width * pil_init_mask.height)
-        init_rgb_pixels = init_rgb_pixels[init_a_pixels > 0]
-        init_rgb_pixels = init_rgb_pixels.reshape(1, init_rgb_pixels.shape[0], init_rgb_pixels.shape[1]) # Filter to just pixels that have any alpha, this is now our histogram
+        if isinstance(init_image, PIL.Image.Image):
```
Comment (Kyle0654) [inline ldm/invoke/generator/inpaint.py:38]: If `init_image` isn't a Pillow image, we won't set `pil_image`, and `sample_to_image` below will break. Should we either assert it's a Pillow image here, or assume it is (possibly adding a type hint to the function definition) and let it fail if not?

Comment (Kyle0654) [reply ldm/invoke/generator/inpaint.py:38]: Same issue with `mask_image` right below this

File: ldm/invoke/generator/inpaint.py (hunk 1 above)
Comment (Kyle0654) [inline ldm/invoke/generator/inpaint.py]: The UI is currently passing RGBA as the mask. Could call it a bug I guess, and make the UI fix it.

**Review by Kyle0654 — CHANGES_REQUESTED** [review]

File: ldm/invoke/generator/inpaint.py (hunk 1 above)
Comment (lstein) [reply ldm/invoke/generator/inpaint.py]: How about we check and convert as necessary?

File: ldm/invoke/generator/inpaint.py (hunk 2 above)
Comment (lstein) [reply ldm/invoke/generator/inpaint.py:38]: Good point. In my version this case would just fall back to the previous behaviour and not try the post processing. I’ll adjust to do the same.

Comment (lstein) [reply ldm/invoke/generator/inpaint.py:38]: I've fixed. In the event that a tensor (or something else) is passed, the post-generation fixup code will not execute and behavior will be as before.

File: ldm/invoke/generator/inpaint.py (hunk 1 above)
Comment (lstein) [reply ldm/invoke/generator/inpaint.py]: Looks for RGB/L vs RGBA and responds appropriately.

**Review by Kyle0654 — APPROVED** [review]

**Pull request merged**